sched: multi level queue support affinity feature

TaiJuWu opened the pull request:
From document, the limit of affinity for multi-level queue is we can't traverse all linked list, but it can be done by using `bitmask` to mask which one bit you have traverse.

This idea is if you already traverse a linked list of specific bit, you can mask it and then continue to traverse another bit.
With `z_priq_dumb_mask_best `, we can traverse all linked list in multi-level queue.

andyross reviewed (commented):
I'm confused on how this is supposed to work?  It needs to be checking the thread cpu_mask field if it wants to do affinity; the bitmasks here are all just the priority queue level flags.

TaiJuWu commented:
> I'm confused on how this is supposed to work? It needs to be checking the thread cpu_mask field if it wants to do affinity; the bitmasks here are all just the priority queue level flags.

checking the thread cpu_mask field is done with `z_priq_dumb_mask_best `; this function is used to traverse priority queue and find the thread can running in current cpu if you set CONFIG_SCHED_CPU_MASK.

From document, the limit of affinity for multi-level queue is we can't traverse all linked list, but it can be done by using `bitmask` to mask which one bit you have traverse.

This idea is if you already traverse a linked list of specific bit, you can mask it and then continue to traverse another bit.
With `z_priq_dumb_mask_best `, we can traverse all linked list in multi-level queue.

peter-mitsis reviewed (dismissed):
As a standalone item, I think this is OK.

That being said, I could have sworn that  you used to have a PR that removed the 32 priority limit to the priq_mq. If my mind is not playing tricks on me, there is going to be some dependencies between them, so extra work might be required.

TaiJuWu commented:
> As a standalone item, I think this is OK.
> 
> That being said, I could have sworn that you used to have a PR that removed the 32 priority limit to the priq_mq. If my mind is not playing tricks on me, there is going to be some dependencies between them, so extra work might be required.

Thanks for review. You are right, I have another PR which is remove 32 priority.
If one has been merged, I will modify another PR to resolve dependencies.

TaiJuWu commented:
@andyross @peter-mitsis Could you take a look?

TaiJuWu committed:
sched: multi level queue support affinity feature

Checking the thread cpu_mask field is done with `z_priq_dumb_mask_best
`; this function is used to traverse priority queue and find the thread
can running in current cpu if you set CONFIG_SCHED_CPU_MASK.

From document, the limit of affinity for multi-level queue is we can't
traverse all linked list, but it can be done by using `bitmask` to mask
which one bit you have traverse.

This idea is if you already traverse a linked list of specific bit, you
can mask it and then continue to traverse another bit.
With `z_priq_dumb_mask_best`, we can traverse all linked list in
multi-level queue.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

cfriedt commented on kernel/include/priority_q.h:
```diff
@@ -199,6 +203,66 @@ static ALWAYS_INLINE struct k_thread *z_priq_dumb_mask_best(sys_dlist_t *pq)
+	struct k_thread *thread = NULL;
+	uint32_t q_idx = 0;
+
+#ifdef CONFIG_64BIT
+	uint64_t bit;
+	uint64_t mask = pq->bitmask[bitmask_idx];
+#else
+	uint32_t bit;
+	uint32_t mask = pq->bitmask[bitmask_idx];
+#endif
+
+	for (; mask != 0; mask &= ~BIT(bit)) {
```
```suggestion
	for (uintptr_t mask = pq->bitmask[bitmask_idx,  bit = 0; mask != 0; mask &= ~BIT(bit)) {
```

cfriedt commented on kernel/include/priority_q.h:
```diff
@@ -199,6 +203,66 @@ static ALWAYS_INLINE struct k_thread *z_priq_dumb_mask_best(sys_dlist_t *pq)
+
+	for (; mask != 0; mask &= ~BIT(bit)) {
+
+#ifdef CONFIG_64BIT
+		bit = u64_count_trailing_zeros(pq->bitmask[bitmask_idx] & mask);
+		q_idx = bitmask_idx * 64 + bit;
+#else
+		bit = u32_count_trailing_zeros(pq->bitmask[bitmask_idx] & mask);
+		q_idx = bitmask_idx * 32 + bit;
+#endif
```
```suggestion
		bit = IS_ENABLED(CONFIG_64BIT) ? u64_count_trailing_zeros(pq->bitmask[bitmask_idx] & mask) : u32_count_trailing_zeros(pq->bitmask[bitmask_idx] & mask);
		q_idx = bitmask_idx * sizeof(mask) * BITS_PER_BYTE + bit;
```

cfriedt commented on kernel/include/priority_q.h:
```diff
@@ -199,6 +203,66 @@ static ALWAYS_INLINE struct k_thread *z_priq_dumb_mask_best(sys_dlist_t *pq)
+
+#if defined(CONFIG_SCHED_MULTIQ)
+
+static ALWAYS_INLINE struct k_thread *_priq_mq_mask_best_idx(struct _priq_mq *pq,
```
Slightly more conventional naming 
```suggestion
static ALWAYS_INLINE struct k_thread *z_priq_mq_best_mask_idx(struct _priq_mq *pq,
```

cfriedt commented on kernel/include/priority_q.h:
```diff
@@ -199,6 +203,66 @@ static ALWAYS_INLINE struct k_thread *z_priq_dumb_mask_best(sys_dlist_t *pq)
+	struct k_thread *thread = NULL;
+
+	for (int i = 0; i < PRIQ_BITMAP_SIZE; ++i) {
+		thread = _priq_mq_mask_best_idx(pq, i);
```
Slightly more conventional naming 
```suggestion
		thread = z_priq_mq_best_mask_idx(pq, i);
```

cfriedt reviewed (commented):
Some minor code refactorings suggested.

Maybe my one concern is misplaced, but to me this algorithm looks `O(n^3)` _at least_.

I'm not 100% sure what the complexity was for the usual non-masked priq version though.

If pinning with the priority queue scheduler goes in, we might want to explicitly mention in documentation somewhere that it might scale very poorly for a large number of threads.

TaiJuWu commented:
> Some minor code refactorings suggested.
> 
> Maybe my one concern is misplaced, but to me this algorithm looks `O(n^3)` _at least_.
> 
> I'm not 100% sure what the complexity was for the usual non-masked priq version though.
> 
> If pinning with the priority queue scheduler goes in, we might want to explicitly mention in documentation somewhere that it might scale very poorly for a large number of threads.

Hi @cfriedt,

 Thanks for your review.

Yes, the time complexity is too large. So I will block here util someone needs this feature or I have any benchmark can proof the feature is useful.
